opencv_ai/splice/tcp.h

```diff
@@ -0,0 +1,100 @@
+#ifndef OPENCV_AI_SPLICE_TCP_H_
+#define OPENCV_AI_SPLICE_TCP_H_
+
+#include <opencv_ai/splice/context.h>
+#include <boost/asio/placeholders.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/connect.hpp>
+#include <boost/asio/read.hpp>
+#include <boost/asio/coroutine.hpp>
+#include <boost/bind.hpp>
+
+#include <glog/logging.h>
+
+namespace opencv_ai {
+using tcp = boost::asio::ip::tcp;
+constexpr const size_t kBufferHeaderLength = 4 + 8 + 4;
+
+class BufferClient {
+ public:
+  BufferClient(Context ctx, std::string host, std::string port,
+               std::function<void(size_t stamp,
+                                  const std::vector<char>& buffer)> handler)
+      : ctx_(ctx),
+        host_(host),
+        port_(port),
+        resolver_(ctx_.GetIoService()),
+        socket_(ctx_.GetIoService()),
+        handler_(std::move(handler)) {
+    CHECK(!!handler_) << "Handler may not be null.";
+    tcp::resolver::query query(tcp::v4(), host, port);
+    resolver_.async_resolve(
+        query, ctx_.Wrap(boost::bind(&BufferClient::HandleResolve, this,
+                                     boost::asio::placeholders::error,
+                                     boost::asio::placeholders::iterator)));
+  }
+
+ private:
+  void HandleResolve(const boost::system::error_code& ec,
+                     tcp::resolver::iterator endpoint_iterator) {
+    CHECK(!ec) << "Could not connect to tcp://" << host_ << ":" << port_
```

Comment:
From the doc: "To make a system of interconnected components crash-only, it must be designed so that components can tolerate the crashes and temporary unavailability of their peers. This means we require strong modularity with relatively impermeable component boundaries, timeout-based communication and lease-based resource allocation, and self-describing requests that carry a time-to-live and information on whether they are idempotent."

I need to read that doc in full again, read about the semantics of the CHECK call, and then I'll be able to have a more intelligent discussion about how to best address network failures in this system.